File: src/server/CloudServer.php
<?php

namespace pocketcloud\cloud\server;

use Closure;
use pocketcloud\cloud\config\Config;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\event\impl\server\ServerStartEvent;
use pocketcloud\cloud\event\impl\server\ServerStopEvent;
use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\client\ServerClientCache;
use pocketcloud\cloud\network\packet\ClientboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\data\NotificationType;
use pocketcloud\cloud\network\packet\data\ServerDisconnectReason;
use pocketcloud\cloud\network\packet\data\VerifyStatus;
use pocketcloud\cloud\network\packet\impl\DisconnectPacket;
use pocketcloud\cloud\network\packet\impl\LanguageSyncPacket;
use pocketcloud\cloud\network\packet\impl\LibrarySyncPacket;
use pocketcloud\cloud\network\packet\impl\MaintenanceListSyncPacket;
use pocketcloud\cloud\network\packet\impl\ModuleSyncPacket;
use pocketcloud\cloud\network\packet\impl\NotificationListSyncPacket;
use pocketcloud\cloud\network\packet\impl\PlayerSyncPacket;
use pocketcloud\cloud\network\packet\impl\ProxyRegisterServerPacket;
use pocketcloud\cloud\network\packet\impl\ServerSyncPacket;
use pocketcloud\cloud\network\packet\impl\TemplateSyncPacket;
use pocketcloud\cloud\player\CloudPlayer;
use pocketcloud\cloud\player\CloudPlayerManager;
use pocketcloud\cloud\server\data\CloudServerData;
use pocketcloud\cloud\server\data\CloudServerStorage;
use pocketcloud\cloud\server\prepare\ServerPreparator;
use pocketcloud\cloud\server\prepare\ServerPrepareEntry;
use pocketcloud\cloud\server\util\CloudServerActionsTrait;
use pocketcloud\cloud\server\util\ServerLogStream;
use pocketcloud\cloud\server\util\ServerStartMethod;
use pocketcloud\cloud\server\util\ServerStatus;
use pocketcloud\cloud\template\Template;
use pocketcloud\cloud\template\TemplateManager;
use pocketcloud\cloud\util\misc\Tickable;
use pocketcloud\cloud\util\misc\Writeable;
use pocketcloud\cloud\util\PathUtils;
use pocketcloud\cloud\util\promise\Promise;
use pocketcloud\cloud\util\Utils;
use Throwable;
use const pocketcloud\CLOUD_PATH;
use const pocketcloud\STATIC_SERVERS_PATH;
use const pocketcloud\TEMP_PATH;

final class CloudServer implements Tickable, Writeable {
    use CloudServerActionsTrait;

    private int $lastCheckTime;
    private ?float $startTime = null;
    private ?int $stopTime = null;
    private ?Config $mainProperties = null;
    private ?ServerStatus $serverStatus;
    private VerifyStatus $verifyStatus;
    private CloudServerStorage $serverStorage;

    public function __construct(
        private readonly int $id,
        private readonly string $serverUuid,
        private readonly string $template,
        private readonly CloudServerData $serverData,
        array $serverStorage = [],
        ?ServerStatus $serverStatus = null
    ) {
        $this->verifyStatus = VerifyStatus::NOT_APPLIED;
        $this->serverStorage = new CloudServerStorage($this, $serverStorage);
        $this->serverStatus = $serverStatus;
    }

    public function tick(int $currentTick): void {
        if ($this->startTime === null) return;
        if ($this->serverStatus === ServerStatus::STARTING) {
            if (($this->startTime + $this->getTemplate()->getTemplateType()->getServerTimeout()) < microtime(true)) {
                $this->handleFailedStart();
            }
        } else if ($this->serverStatus?->isOnline()) {
            if (!$this->checkAlive()) {
                $this->handleTimeout();
            }
        } else if ($this->serverStatus === ServerStatus::STOPPING) {
            if (($this->getStopTime() + 10) <= time()) {
                $this->handleStopTimeout();
            }
        } else if ($this->serverStatus === ServerStatus::OFFLINE) {
            $this->remove();
            if ($this->checkForCrash()) $this->killProcess();
            $this->deleteTmpDir();
        }
    }

    public function prepare(): Promise {
        $promise = new Promise();
        ServerPreparator::getInstance()->submitEntry($this, ServerPrepareEntry::fromServer($this), fn() => $promise->resolve($this), fn(?Throwable $e) => $promise->reject([$this->getName(), $e]));
        return $promise;
    }

    public function start(): void {
        $this->setServerStatus(ServerStatus::STARTING);
        new ServerStartEvent($this)->call();
        CloudLogger::get()->info("§aStarting §b{}§r...", $this);
        NotificationType::SERVER_STARTING->notify(["server" => $this->getName()]);

        ServerStartMethod::current()?->startServer($this)->then(function (?int $tmpPid): void {
            $this->startTime = microtime(true);
            $this->serverData->setTempProcessId($tmpPid);
        })->failure(function (): void {
            CloudLogger::get()->warn("Failed to start server §b{}§8, §rcould not create the process...", $this);
            NotificationType::SERVER_START_FAILED->notify(["server" => $this->getName(), "reason" => "Failed to create process"]);
            $this->remove();
            $this->deleteTmpDir();
        });
    }

    public function stop(bool $force = false): void {
        new ServerStopEvent($this, $force)->call();
        CloudLogger::get()->info("§cStopping §b{}§r...{}", $this->getName(), $force ? " §8(§cforcefully!§8)" : "");
        NotificationType::SERVER_STOPPING->notify(["server" => $this->getName()]);
        $this->setServerStatus(ServerStatus::STOPPING);
        $this->setStopTime(time());

        if ($force) {
            $this->setServerStatus(ServerStatus::OFFLINE);
            $this->killProcess();
            $this->remove();
            $this->checkForCrash();
            $this->deleteTmpDir();
        } else {
            DisconnectPacket::create(ServerDisconnectReason::SERVER_SHUTDOWN)->sendPacket($this);
        }
    }

    public function sync(): void {
        $packets = [
            LanguageSyncPacket::fromLanguage(), LibrarySyncPacket::fromLibraries(),
            ModuleSyncPacket::fromModuleCache(), MaintenanceListSyncPacket::fromMaintenanceListCache(), NotificationListSyncPacket::fromNotificationListCache()
        ];

        foreach (TemplateManager::getInstance()->getAll() as $template) $packets[] = TemplateSyncPacket::create($template, false);
        foreach (CloudServerManager::getInstance()->getAll() as $server) {
            if ($server->getServerStatus() === null) continue;
            $packets[] = ServerSyncPacket::create($server, false);
            if ($this->getTemplate()->getTemplateType()->isProxy() && $server->getTemplate()->getTemplateType()->isServer()) $packets[] = ProxyRegisterServerPacket::create($server->getName(), $server->getServerData()->getPort());
        }

        foreach (CloudPlayerManager::getInstance()->getAll() as $player) $packets[] = PlayerSyncPacket::create($player, false);

        foreach ($packets as $packet) $this->sendPacket($packet);
    }

    public function openLogStream(): ServerLogStream {
        return new ServerLogStream($this);
    }

    /**
     * @param CloudPacket $packet
     * @param int $ticks delay in ticks (20 = 1s)
     * @param Closure(ServerClient $client, CloudPacket $packet, bool $success): void|null $onSend
     * @return void
     */
    public function sendDelayedPacket(CloudPacket $packet, int $ticks, ?Closure $onSend = null): void {
        $this->getServerClient()?->sendDelayedPacket($packet, $ticks, $onSend);
    }

    public function sendPacket(ClientboundPacket $packet): bool {
        return $this->getServerClient()?->sendPacket($packet) ?? false;
    }

    public function retrieveLogs(): ?array {
        if (@file_exists($path = $this->getLogFilePath())) {
            return explode("\n", file_get_contents($path));
        }

        return null;
    }

    public function setServerStatus(ServerStatus $serverStatus): void {
        $this->serverStatus = $serverStatus;
        ServerSyncPacket::create($this, false)->broadcastPacket();
    }

    public function setLastCheckTime(float $lastCheckTime): void {
        $this->lastCheckTime = $lastCheckTime;
    }

    public function setStopTime(float $stopTime): void {
        $this->stopTime = $stopTime;
    }

    public function setVerifyStatus(VerifyStatus $verifyStatus): void {
        $this->verifyStatus = $verifyStatus;
    }

    public function checkAlive(): bool {
        $timeout = $this->getTemplate()->getTemplateType()->getServerTimeout();
        if ((microtime(true) - $this->startTime) < $timeout) return true;
        if (!isset($this->lastCheckTime)) return false;
        if ((time() - $this->lastCheckTime) < $timeout) return true;
        return false;
    }

    public function getServerClient(): ?ServerClient {
        return ServerClientCache::getInstance()->get($this);
    }

    public function getPlayer(string $name): ?CloudPlayer {
        return array_find($this->getPlayers(), fn(CloudPlayer $player) => $player->getName() == $name ||
            $player->getUniqueId() == $name ||
            $player->getXboxUserId() == $name
        );
    }

    /** @return array<CloudPlayer> */
    public function getPlayers(): array {
        return CloudPlayerManager::getInstance()->getAll($this);
    }

    public function getPlayerCount(): int {
        return count($this->getPlayers());
    }

    public function getPath(): string {
        if ($this->getTemplate()?->isStatic()) return PathUtils::join(STATIC_SERVERS_PATH, $this->getName()) . "/";
        return PathUtils::join(TEMP_PATH, $this->serverUuid) . "/";
    }

    public function getLogFilePath(): string {
        $basePath = $this->getPath();
        $logFile = $this->getTemplate()->getTemplateType()->getRelativeLogFileLocation();
        return $basePath . $logFile;
    }

    public function getProperties(): Config {
        return $this->mainProperties ??= new Config(PathUtils::join($this->getPath(), $this->getTemplate()->getTemplateType()->getMainConfigurationFile()));
    }

    public function getServerUuid(): string {
        return $this->serverUuid;
    }

    public function getName(): string {
        return $this->template . "-" . $this->id;
    }

    public function getId(): int {
        return $this->id;
    }

    public function getTemplate(): Template {
        return TemplateManager::getInstance()->get($this->template);
    }

    public function getTemplateName(): string {
        return $this->template;
    }

    public function getServerData(): CloudServerData {
        return $this->serverData;
    }

    public function getServerStatus(): ?ServerStatus {
        return $this->serverStatus;
    }

    public function getStartTime(): ?float {
        return $this->startTime;
    }

    public function getLastCheckTime(): float {
        return $this->lastCheckTime;
    }

    public function getServerStorage(): CloudServerStorage {
        return $this->serverStorage;
    }

    public function getStopTime(): ?float {
        return $this->stopTime;
    }

    public function getVerifyStatus(): VerifyStatus {
        return $this->verifyStatus;
    }

    public function write(): array {
        return [
            "name" => $this->getName(),
            "uuid" => $this->serverUuid,
            "id" => $this->id,
            "template" => $this->template,
            "serverStatus" => $this->serverStatus?->getName(),
            "port" => $this->serverData->getPort(),
            "players" => $this->getPlayerCount(),
            "maxPlayers" => $this->serverData->getMaxPlayers(),
            "processId" => $this->serverData->getProcessId(),
            "tps" => $this->serverData->getTps(),
            "avgTps" => $this->serverData->getAvgTps(),
            "memoryUsage" => $this->serverData->getMemoryUsage(),
            "memoryPeak" => $this->serverData->getMemoryPeak(),
            "memoryLimit" => $this->serverData->getMemoryLimit(),
            "cpuUsage" => $this->serverData->getCpuUsage(),
            "internalStorage" => $this->serverStorage->getAll()
        ];
    }

    public function __toString(): string {
        return "§b" . $this->getName() . " §8[§ruuid=" . $this->serverUuid . " path=" . trim(str_replace(CLOUD_PATH, "", $this->getPath()), "/") . " port=" . $this->getServerData()->getPort() . "§8]§r";
    }

    public static function read(array $data): ?self {
        if (!Utils::containKeys($data, "name", "uuid", "id", "template", "port", "maxPlayers", "processId", "serverStatus")) return null;
        if (($template = TemplateManager::getInstance()->get($data["template"])) === null) return null;
        return new CloudServer(
            intval($data["id"]),
            $data["uuid"],
            $template,
            new CloudServerData($data["name"], intval($data["port"]), intval($data["maxPlayers"]), ($data["processId"] === null ? null : intval($data["processId"]))),
            $data["internalStorage"] ?? [],
            ServerStatus::fromName($data["serverStatus"]) ?? ServerStatus::ONLINE
        );
    }
}

File: src/main/java/de/pocketcloud/cloud/server/CloudServer.java
package de.pocketcloud.cloud.server;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.config.Config;
import de.pocketcloud.cloud.config.exception.UnsupportedFileExtensionException;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.console.log.def.PrefixedLogger;
import de.pocketcloud.cloud.event.impl.server.ServerPrepareEvent;
import de.pocketcloud.cloud.event.impl.server.ServerStartEvent;
import de.pocketcloud.cloud.event.impl.server.ServerStopEvent;
import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.client.ServerClientCache;
import de.pocketcloud.cloud.network.packet.ClientboundPacket;
import de.pocketcloud.cloud.network.packet.impl.*;
import de.pocketcloud.cloud.network.packet.type.NotificationType;
import de.pocketcloud.cloud.network.packet.type.ServerDisconnectReason;
import de.pocketcloud.cloud.network.packet.type.VerificationStatus;
import de.pocketcloud.cloud.player.CloudPlayer;
import de.pocketcloud.cloud.player.CloudPlayerManager;
import de.pocketcloud.cloud.server.config.IServerProperties;
import de.pocketcloud.cloud.server.config.ServerPropertiesGenerator;
import de.pocketcloud.cloud.server.util.CloudServerData;
import de.pocketcloud.cloud.server.util.CloudServerStorage;
import de.pocketcloud.cloud.server.util.ServerStartMethods;
import de.pocketcloud.cloud.server.util.ServerStatus;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.TemplateManager;
import de.pocketcloud.cloud.template.TemplateType;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.cloud.tick.Tickable;
import de.pocketcloud.cloud.util.*;
import de.pocketcloud.cloud.util.concurrent.Promise;
import de.pocketcloud.cloud.util.mapper.MapperUtils;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.nio.file.attribute.FileTime;
import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.CompletableFuture;

import static de.pocketcloud.cloud.util.FileUtils.IO_EXECUTOR;

@Getter
@Accessors(fluent = true)
public final class CloudServer implements Tickable, Writable<Map<String, Object>>, FilterableObject {

    private final int id;
    private final UUID uuid;
    private final String templateName;
    private final CloudServerData serverData;
    private final CloudServerStorage storage;

    private boolean pidLookupDone = false;
    private long lastPidLookupTime = 0;
    private int lastPidLookupCounter = 0;

    private final PrefixedLogger logger;
    @Setter
    private VerificationStatus verificationStatus = VerificationStatus.PENDING;
    private ServerStatus status = ServerStatus.PENDING;
    @Setter
    private Long lastKeepAlive = null;
    private Long startTime = null;
    @Setter
    private Long verifiedTime = null;
    private Long stopTime = null;
    private Config mainProperties = null;

    public CloudServer(int id, UUID uuid, String templateName, CloudServerData serverData) {
        this.id = id;
        this.uuid = uuid;
        this.templateName = templateName;
        this.serverData = serverData;
        this.storage = new CloudServerStorage(uuid);
        this.logger = new PrefixedLogger(PocketCloud.instance().logger(), "§8[§b" + name() + "§r§8]");
    }

    public CloudServer(int id, UUID uuid, String templateName, CloudServerData serverData, ServerStatus status) {
        this(id, uuid, templateName, serverData);
        this.status = status;
    }

    public CloudServer(int id, UUID uuid, String templateName, CloudServerData serverData, Map<String, Object> storage) {
        this(id, uuid, templateName, serverData);
        this.storage.setAll(storage);
    }

    @Override
    public void tick(long currentTick) {
        if (startTime == null) return;
        if (status == ServerStatus.STARTING) {
            if (!pidLookupDone && serverData.processId() == null && (System.currentTimeMillis() - lastPidLookupTime) >= 1500 && lastPidLookupCounter < 5) {
                lastPidLookupTime = System.currentTimeMillis();
                lastPidLookupCounter++;
                ServerStartMethods.current().lookupPid(this).thenSuccess(pid -> {
                    if (pid.isPresent()) {
                        pidLookupDone = true;
                        serverData.processId(pid.get());
                    }
                });
            }

            if ((startTime + (template().templateType().timeout() * 1000L)) < System.currentTimeMillis()) {
                CloudServersHandler.handleStartFailure(this, null, false);
            }
        }
    }

    public Promise<Void> prepare() {
        new ServerPrepareEvent(this).call();
        return Promise.runAsync(() -> {
            Path logFileLocation = logFilePath();
            if (Files.exists(logFileLocation)) {
                Path logArchivePath = template().path().resolve("cloud_log_archive");
                FileUtils.createDir(logArchivePath);

                try {
                    FileTime ctime = Files.readAttributes(logFileLocation, BasicFileAttributes.class).creationTime();
                    String timestamp = DateTimeFormatter
                            .ofPattern("yyyy-MM-dd_HH:mm:ss.SSS_z")
                            .withZone(ZoneId.systemDefault())
                            .format(ctime.toInstant());
                    String archiveFileName = timestamp + "_" + logFileLocation.getFileName() + ".log";
                    Files.copy(logFileLocation, logArchivePath.resolve(archiveFileName), StandardCopyOption.REPLACE_EXISTING);
                } catch (IOException e) {
                    CloudLogger.get().exception("Failed to archive log file", e);
                }

                FileUtils.unlinkFile(logFileLocation);
            }

            if (Files.exists(path()) && !template().settings().isStaticServers()) {
                FileUtils.removeDirectory(path());
            }

            Path bridgePath = template().serverSoftware().bridge().sourceFile();
            Path bridgeDstPath = path().resolve(template().serverSoftware().bridge().relativeServerPath());
            if (!Files.isDirectory(bridgeDstPath.getParent())) FileUtils.createDir(bridgeDstPath.getParent());
            if (Files.isRegularFile(bridgePath)) {
                try {
                    Files.copy(bridgePath, bridgeDstPath, StandardCopyOption.REPLACE_EXISTING);
                } catch (IOException e) {
                    CloudLogger.get().exception("Failed to copy bridge", e);
                }
            }

            Path globalTemplatePath = template().templateType().globalTemplatePath();
            FileUtils.copyDirectory(globalTemplatePath, path(), Set.of());

            boolean copyFromSources = template().settings().isAlwaysCopyToStaticServers() || !template().settings().isStaticServers();
            if (copyFromSources) {
                for (ServerGroup group : template().parentGroups()) {
                    FileUtils.copyDirectory(group.path(), path(), Set.of());
                }

                FileUtils.copyDirectory(template().path(), path(), Set.of("cloud_log_archive/"));
            }

            List<IServerProperties> propertiesData = ServerPropertiesGenerator.instance().getAll(template().serverSoftware());
            for (var properties : propertiesData) {
                Path filePath = path().resolve(properties.getFileName());
                if (!copyFromSources) {
                    try {
                        Files.copy(
                                globalTemplatePath.resolve(properties.getFileName()),
                                filePath,
                                StandardCopyOption.REPLACE_EXISTING
                        );
                    } catch (IOException e) {
                        CloudLogger.get().exception("Failed to copy properties", e);
                    }
                }

                Map<String, Object> replacements = properties.replacePlaceholders(this);
                String content = FileUtils.fileGetContents(filePath);
                for (Map.Entry<String, Object> entry : replacements.entrySet()) {
                    content = content.replace(entry.getKey(), entry.getValue().toString());
                }

                FileUtils.filePutContents(filePath, content);
            }

            FileUtils.unlinkFile(logFileLocation);
        }, IO_EXECUTOR);
    }

    public CloudServer start() {
        startTime = System.currentTimeMillis();
        setStatus(ServerStatus.STARTING);
        new ServerStartEvent(this).call();
        CloudLogger.get().info("§aStarting §b{} §8[§ruuid={}, path={}, port={}§8]§r...", name(), uuid.toString(), path().toString(), serverData.port());
        NotificationType.SERVER_STARTING.notify(Map.of("server", name()));
        return this;
    }

    public void boot() {
        ServerStartMethods.current().start(new CloudServer[]{this})
                .thenSuccess(m -> CloudServersHandler.handleStartSuccess(this, m.getOrDefault(name(), null)))
                .failure(t -> CloudServersHandler.handleStartFailure(this, t, true));
    }

    public void stop(boolean force) {
        new ServerStopEvent(this, force).call();
        NotificationType.SERVER_STOPPING.notify(Map.of("server", name()));
        stopTime = System.currentTimeMillis();

        if (force) {
            CloudLogger.get().info("§cStopped §b{} §rforcefully.", name());
            setStatus(ServerStatus.OFFLINE);
            kill();
            remove();
            //todo checkforCrash
            deleteTmpDir();
        } else {
            CloudLogger.get().info("§cStopping §b{}§r...", name());
            setStatus(ServerStatus.STOPPING);
            sendPacket(DisconnectPacket.create(ServerDisconnectReason.SERVER_SHUTDOWN));
        }
    }

    public void stop() {
        stop(false);
    }

    public void saveAndDeleteLogFiles() {
        Path logFileLocation = logFilePath();
        Path logArchivePath = template().path().resolve("cloud_log_archive");
        if (Files.exists(logFileLocation)) {
            if (!Files.isDirectory(logArchivePath)) {
                try {
                    Files.createDirectory(logArchivePath);
                } catch (IOException e) {
                    CloudLogger.get().exception("Failed to create log archive directory", e);
                }
            }

            String formatted = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH:mm:ss.SSS_z")
                    .format(Instant.ofEpochMilli((long) Math.floor(startTime))
                            .atZone(ZoneId.systemDefault()));

            try {
                Files.copy(logFileLocation, logArchivePath.resolve(formatted + "_" + logFileLocation.getFileName().toString() + ".log"));
            } catch (IOException e) {
                CloudLogger.get().exception("Failed to copy log file", e);
            }
        }
    }

    public Promise<Void> save() {
        return Promise.supplyAsync(() -> {
            for (String file : template().serverSoftware().config().savableFiles()) {
                Path filePath = path().resolve(file);
                Path dstPath = template().path().resolve(file);
                if (Files.isRegularFile(filePath)) {
                    try {
                        Files.copy(filePath, dstPath);
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                } else if (Files.isDirectory(filePath)) {
                    FileUtils.copyDirectory(filePath, dstPath, Set.of());
                }
            }

            return null;
        }, IO_EXECUTOR);
    }

    public void deleteTmpDir() {
        saveAndDeleteLogFiles();
        if (!template().settings().isStaticServers()) {
            CloudLogger.get().debug("Removed server directory {}", path().toString());
            FileUtils.removeDirectory(path());
        }
    }

    public void remove() {
        CloudServerManager.instance().remove(this);
        ServerClientCache.instance().remove(this);

        if (template().templateType() == TemplateType.SERVER) {
            if (template().settings().isSaveOnShutdown()) save();
            removeFromProxies();
        }
    }

    public void kill() {
        if (serverData.processId() != null) ProcessUtils.kill(serverData.processId(), true);
    }

    public void addToProxies() {
        ProxyRegisterServerPacket.create(this).broadcastPacket(TemplateType.SERVER);
    }

    public void removeFromProxies() {
        ProxyUnregisterServerPacket.create(this).broadcastPacket(TemplateType.SERVER);
    }

    public void sync() {
        List<ClientboundPacket> syncPackets = new ArrayList<>(List.of(
                LanguageSyncPacket.fromLanguage(),
                LibrarySyncPacket.fromLibraries(this),
                ModuleSyncPacket.fromModuleCache(),
                MaintenanceListSyncPacket.fromMaintenanceListCache(),
                NotificationListSyncPacket.fromNotificationListCache(),
                BulkSyncPacket.generate()
        ));

        if (template().templateType().isProxy()) {
            for (CloudServer subServer : CloudServerManager.instance().getAll(TemplateType.SERVER)) {
                if (subServer.status().isOnline()) {
                    syncPackets.add(ProxyRegisterServerPacket.create(subServer));
                }
            }
        }

        for (ClientboundPacket packet : syncPackets) {
            sendPacket(packet);
        }
    }

    public CompletableFuture<Void> sendPacket(ClientboundPacket packet) {
        if (client().isEmpty()) return CompletableFuture.failedFuture(new RuntimeException("Client is empty"));
        return client().get().sendPacket(packet);
    }

    public CompletableFuture<Void> sendDelayedPacket(ClientboundPacket packet, long delayMs) {
        if (client().isEmpty()) return CompletableFuture.failedFuture(new RuntimeException("Client is empty"));
        return client().get().sendDelayedPacket(packet, delayMs);
    }

    public void setStatus(ServerStatus status) {
        this.status = status;
        ServerSyncPacket.create(this, false).broadcastPacket();
    }

    public String name() {
        return templateName + "-" + id;
    }

    public Template template() {
        return TemplateManager.instance().get(templateName).orElseThrow(() -> new RuntimeException("Template null? This should not happen"));
    }

    public Path path() {
        if (template().settings().isStaticServers()) return PocketCloudPaths.storage().staticServers().with(name()).asPath();
        return PocketCloudPaths.tmp().with(uuid.toString()).asPath();
    }

    public Path logFilePath() {
        return path().resolve(template().serverSoftware().config().relativeLogFileLocation());
    }

    public Config properties() {
        if (mainProperties == null) {
            try {
                mainProperties = new Config(path().resolve(template().serverSoftware().config().mainConfigurationFile()));
            } catch (IOException | UnsupportedFileExtensionException e) {
                throw new RuntimeException(e);
            }
        }

        return mainProperties;
    }

    public Optional<ServerClient> client() {
        return ServerClientCache.instance().get(this);
    }

    public List<CloudPlayer> players() {
        return CloudPlayerManager.instance().getAll(this);
    }

    public int playerCount() {
        return players().size();
    }

    @Override
    public Map<String, Object> write() {
        return MapperUtils.toMap(this);
    }

    public static CloudServer read(Map<String, Object> data) {
        return MapperUtils.fromMap(data, CloudServer.class);
    }
}